src/nfa/nfa.cc

```diff
@@ -312,6 +312,136 @@ struct StackLevel {
+                if (this->post_it != this->post_end) {
+                    this->targets_it = this->post_it->cbegin();
+                    this->targets_end = this->post_it->cend();
+                }
+            } 
+        }
+    }
+
+    State get_act_succ() {
+        align_succ();
+        return *(this->targets_it);
+    }
+
+    void move_next_succ() {
+        if(this->post_it == this->post_end) {
+            return;
+        }
+        this->targets_it++;
+    }
+
+    bool is_end() {
+        align_succ();
+        return this->post_it == this->post_end;
+    }
+};
+
+BoolVector Nfa::get_useful_states_tarjan() const {
+    BoolVector reached_and_reaching(this->size(),false); 
+    std::vector<NodeData> node_info(this->size());
+    std::deque<State> program_stack;
+    std::deque<State> tarjan_stack;
+    unsigned long cnt = 0;
+
+    for(const State& q0 : initial) {
+        program_stack.push_back(q0);
+    }
+
+    while(!program_stack.empty()) {
+        State act_state = program_stack.back();
+        NodeData& act_state_data = node_info[act_state];
+        // node has not been initialized yet --> corresponds to the first call of strongconnect(act_state)
+        if(!act_state_data.initilized) {
+            // initialize node
+            act_state_data = NodeData(act_state, this->delta, cnt++);
+            tarjan_stack.push_back(act_state);
+            if(this->final.contains(act_state)) {
+                reached_and_reaching[act_state] = true;
+            }
+        } else { // return from the recursive call
+            State act_succ = act_state_data.get_act_succ();
+            act_state_data.lowlink = std::min(act_state_data.lowlink, node_info[act_succ].lowlink);
+            // act_succ is the state that cased the recursive call. Move to another successor.
+            act_state_data.move_next_succ();
+        }
+
+        // iterate through outgoing edges
+        State next_state;
+        bool rec_call = false;
+        while(!act_state_data.is_end()) {
+            next_state = act_state_data.get_act_succ();
+            // if successor is useful, act_state is useful as well
+            if(reached_and_reaching[next_state]) {
+                reached_and_reaching[act_state] = true;
+            }
+            if(!node_info[next_state].initilized) { // recursive call
+                program_stack.push_back(next_state);
+                rec_call = true;
+                break;
+            } else if(node_info[next_state].on_stack) {
+                act_state_data.lowlink = std::min(act_state_data.lowlink, node_info[next_state].index);
+            }
+            act_state_data.move_next_succ();
+        }
+        if(rec_call) continue;
+
+        // check if we have the root of a SCC
+        if(act_state_data.lowlink == act_state_data.index) {
+            State st;
+            bool final_scc = false;
+            std::vector<State> scc;
+            do {
+                st = tarjan_stack.back();
```

Comment:
I wonder, can `tarjan_stack` be empty at this line? Have you tried it on some extreme cases? Empty Nfa or maybe One State Nfa?

Reply:
Should not, in theory. That's all I can say. I can put more tests there.

Reply:
Can you please write a theoretical proof, that this cannot happen? That should suffice. (that is just joke). You are much smarter than me, so I will trust you.

Reply:
I guess you will regret 😄

Reply:
Or give it to Adam, he can find failing corner cases while taking a nap.